README.md
rails_app Cookbook
==================
[![Build Status](https://travis-ci.org/jhx/cookbook-rails_app.png?branch=master)](https://travis-ci.org/jhx/cookbook-rails_app)
[![Dependency Status](https://gemnasium.com/jhx/cookbook-rails_app.png)](https://gemnasium.com/jhx/cookbook-rails_app)

Installs support packages and symlinks for the Rails app.


Requirements
------------
### Cookbooks
The following cookbooks are direct dependencies because they're used for common "default" functionality.

- `chef-sugar`
- `cron`
- `database`
- `logrotate`
- `mysql`
- `rvm`
- `yum`

### Platforms
The following platforms are supported and tested under Test Kitchen:

- CentosOS 5.10, 6.5

Other RHEL family distributions are assumed to work. See [TESTING](TESTING.md) for information about running tests in Opscode's Test Kitchen.


Attributes
----------
Refer to [`attributes/default.rb`](attributes/default.rb) for default values.

- `node['rails_app']['name']` - name of Rails app
- `node['rails_app']['rails']['version]` - version of Rails gem (not currently used in this cookbook)
- `node['rails_app']['stages']` - options hash to configure database


Recipes
-------
This cookbook provides one main recipe for configuring a node.

- `db.rb` - *Use this recipe* to install and configure Rails app database server.
- `web.rb` - *Use this recipe* to install and configure Rails app webserver.

### db
This recipe creates and configures databases and database users.

### default
This recipe is a no-op.

### web
This recipe installs required packages and symlinks for the Rails app webserver.


Usage
-----
On client nodes, use the following recipes:

````javascript
{ "run_list": ["recipe[rails_app::db]", "recipe[rails_app::web]"] }
````

The following are the key items achieved by this cookbook:

- creates required databases and database users
- installs YUM EPEL repository
- installs required support packages
- creates required symlinks for Rails app webserver
- installs `logrotate` configuration file
- configures SSH access for deploy user


License & Authors
-----------------
- Author:: Doc Walker (<doc.walker@jameshardie.com>)

````text
Copyright 2013-2014, James Hardie Building Products, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
````


CHANGELOG.md
CHANGELOG
=========

0.4.0 / 2014-02-09
------------------

- `NEW` add emeril gem to publish cookbook | [#40][]
- `FIX` add missing require 'spec_helper' | [#39][]
- `NEW` add [CHANGELOG](CHANGELOG.md) | [#37][]
- `IMPROVE` update [README](README.md) | [#35][] [#36][]
- `IMPROVE` remove deprecated functionality | [#34][]
- `NEW` add [TESTING](TESTING.md) | [#32][]
- `NEW` add integration tests (via test-kitchen/serverspec) | [#32][]
- `IMPROVE` require rspec expect syntax | [#26][]
- `IMPROVE` update .gitignore | [#23][]
- `NEW` add [CONTRIBUTING](CONTRIBUTING.md) | [#18][]
- `NEW` add ruby lint checks (via rubocop), fix flagged items | [#17][] [#22][] [#24][]
- `NEW` add unit tests (via chefspec) | [#15][] [#16][]
- `NEW` add syntax checks (via knife) | [#14][]
- `NEW` add cookbook lint checks (via foodcritic), fix flagged items | [#12][] [#13][]
- `NEW` add specification files for ruby version | [#11][]


0.3.5 / 2013-08-14
------------------

- `IMPROVE` remove deprecated `cron` job


0.3.4 / 2013-07-08
------------------

- `IMPROVE` add fallback password for all database users


0.3.3 / 2013-07-08
------------------

- `FIX` change fallback password to string (was symbol)


0.3.2 / 2013-07-08
------------------

- `IMPROVE` add fallback password for `insql` database user


0.3.1 / 2013-07-08
------------------

- `IMPROVE` add database user `insql` to all accessible domains


0.3.0 / 2013-07-03
------------------

- `NEW` add database user `insql` for delay app


0.2.0 / 2013-02-13
------------------

- `NEW` add support for remote deploys


0.1.1 / 2013-01-26
------------------

- `IMPROVE` change guard in `rails_app::web` recipe
- `IMPROVE` update spec to use `logrotate_` cookbook


0.1.0 / 2013-01-26
------------------

- `NEW` initial release


Legend
------

- `BREAK`   - breaking changes
- `FIX`     - bug fix
- `IMPROVE` - improvement of existing feature
- `NEW`     - new feature

<!--- The following link definition list is generated by PimpMyChangelog --->
[#11]: https://github.com/jhx/cookbook-rails_app/issues/11
[#12]: https://github.com/jhx/cookbook-rails_app/issues/12
[#13]: https://github.com/jhx/cookbook-rails_app/issues/13
[#14]: https://github.com/jhx/cookbook-rails_app/issues/14
[#15]: https://github.com/jhx/cookbook-rails_app/issues/15
[#16]: https://github.com/jhx/cookbook-rails_app/issues/16
[#17]: https://github.com/jhx/cookbook-rails_app/issues/17
[#18]: https://github.com/jhx/cookbook-rails_app/issues/18
[#22]: https://github.com/jhx/cookbook-rails_app/issues/22
[#23]: https://github.com/jhx/cookbook-rails_app/issues/23
[#24]: https://github.com/jhx/cookbook-rails_app/issues/24
[#26]: https://github.com/jhx/cookbook-rails_app/issues/26
[#32]: https://github.com/jhx/cookbook-rails_app/issues/32
[#34]: https://github.com/jhx/cookbook-rails_app/issues/34
[#35]: https://github.com/jhx/cookbook-rails_app/issues/35
[#36]: https://github.com/jhx/cookbook-rails_app/issues/36
[#37]: https://github.com/jhx/cookbook-rails_app/issues/37
[#39]: https://github.com/jhx/cookbook-rails_app/issues/39
[#40]: https://github.com/jhx/cookbook-rails_app/issues/40


spec/db_spec.rb
# coding: utf-8
require 'spec_helper'

describe 'rails_app::db' do
  before do
    Chef::Sugar::DataBag.stub(:encrypted_data_bag_item).and_return(
      'matrix' => 'matrix_password',
      'matrix_staging' => 'matrix_staging_password',
      'mysql' => {
        'root' => 'root_password'
      },
      'mysqladmin' => 'mysqladmin_password',
      'wwuser' => 'wwuser_password'
    )
  end # before

  let(:chef_run) do
    ChefSpec::Runner.new do |node|
      # create a new environment
      env = Chef::Environment.new
      env.name 'qa'

      # stub the node to return this environment
      node.stub(:chef_environment).and_return(env.name)

      # stub any calls to Environment.load to return this environment
      Chef::Environment.stub(:load).and_return(env)

      # override cookbook attributes
      # node.set['key'] = some_value

      # required for build-essential cookbook on travis-ci
      node.set['platform_family'] = 'rhel'
    end.converge(described_recipe)
  end # let

  it 'includes recipe mysql::ruby' do
    expect(chef_run).to include_recipe('mysql::ruby')
  end # it

  it 'includes recipe chef-sugar' do
    expect(chef_run).to include_recipe('chef-sugar')
  end # it

  %w(localhost %).each do |domain|
    it "creates user mysqladmin@#{domain}" do
      pending "creates user mysqladmin@#{domain}"
    end # it
  end # .each

  %w(matrix_production matrix_staging).each do |database|
    it "creates database #{database}" do
      pending "creates database #{database}"
    end # it
  end # .each

  it 'grants privileges to user matrix' do
    pending 'grants privileges to user matrix'
  end # it

  it 'grants privileges to user matrix_staging' do
    pending 'grants privileges to user matrix_staging'
  end # it

  it 'grants privileges to user wwuser' do
    pending 'grants privileges to user wwuser'
  end # it

  it 'drops database test' do
    pending 'drops database test'
  end # it

end # describe


spec/web_spec.rb
# coding: utf-8
require 'spec_helper'

describe 'rails_app::web' do
  before do
    Chef::Sugar::DataBag.stub(:encrypted_data_bag_item).and_return(
      'stage1_db_username' => 'stage1_db_password',
      'stage2_db_username' => 'stage2_db_password'
    )
  end # before

  let(:chef_run) do
    ChefSpec::Runner.new do |node|
      # create a new environment
      env = Chef::Environment.new
      env.name 'qa'

      # stub the node to return this environment
      node.stub(:chef_environment).and_return(env.name)

      # stub any calls to Environment.load to return this environment
      Chef::Environment.stub(:load).and_return(env)

      # override cookbook attributes
      node.set['file']['header'] = 'node.file.header'
      node.set['rails_app']['name'] = 'whiz_bang_app'
      node.set['rails_app']['stages'] = [
        {
          'name'                    => 'stage1',
          'codename'                => 'st1',
          'db_username'             => 'stage1_db_user'
        },
        {
          'name'                    => 'stage2',
          'codename'                => 'st2',
          'db_username'             => 'stage2_db_user'
        }
      ]

      # required for build-essential cookbook on travis-ci
      node.set['platform_family'] = 'rhel'
    end.converge(described_recipe)
  end # let

  let(:node) { chef_run.node }

  it 'creates directory /var/www/html' do
    dir = '/var/www/html'
    expect(chef_run).to create_directory(dir)
      .with(:owner => 'root', :group => 'root')
  end # it

  it 'creates user jeeves' do
    expect(chef_run).to create_user('jeeves')
  end # it

  it 'creates multiple symlinks' do
    node['rails_app']['stages'].each do |item|
      link = "/var/www/html/#{item['codename']}"
      expect(chef_run).to create_link(link)
        .with(:owner => 'root', :group => 'root')
    end # .each
  end # it

  it 'includes recipe logrotate_::var_www_apps' do
    expect(chef_run).to include_recipe('logrotate')
  end # it

  describe '/home/jeeves/.ssh' do
    it 'creates directory' do
      expect(chef_run).to create_directory(subject)
        .with(:owner => 'jeeves', :group => 'jeeves', :mode => '0700')
    end # it
  end # describe

  describe '/home/jeeves/.ssh/config' do
    it 'creates template' do
      expect(chef_run).to create_template(subject)
        .with(:owner => 'jeeves', :group => 'jeeves', :mode => '0644')
    end # it

    it 'contains expected content' do
      expect(chef_run).to render_file(subject)
        .with_content('node.file.header')
    end # it
  end # describe

end # describe


spec/spec_helper.rb
# coding: utf-8
require 'chef/sugar'
require 'chefspec'
require 'chefspec/berkshelf'
require 'chefspec/deprecations'


recipes/db.rb
# coding: utf-8
#
# Cookbook Name:: rails_app
# Recipe:: db
#
# Author:: Doc Walker (<doc.walker@jameshardie.com>)
#
# Copyright 2012-2013, James Hardie Building Products, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

include_recipe 'mysql::ruby'
include_recipe 'chef-sugar'

# retrieve contents of encrypted data bag (refer to chef-repo/ENCRYPTED.md)
secret = encrypted_data_bag_item(:encrypted, node.chef_environment)

# establish database server connection parameters
connection_info = {
  :host => 'localhost',
  :username => 'root',
  :password => node['mysql']['server_root_password'] ||
    secret['mysql']['root']
}


# grant privileges to user 'mysqladmin' for super-user access
%w(localhost %).each do |domain|
  # mysql_database 'custom query' do
  #   connection connection_info
  #   sql <<-EOF
  #   GRANT ALL PRIVILEGES
  #   ON *.*
  #   TO `mysqladmin`@`#{domain}`
  #   IDENTIFIED BY PASSWORD '#{secret["mysqladmin"]}'
  #   WITH GRANT OPTION;
  #   EOF
  #   action :query
  #   only_if { secret['mysqladmin'] }
  # end
  mysql_database_user 'mysqladmin' do
    connection connection_info
    password secret['mysqladmin'] || 'missing_password'
    host domain
    with_option ['GRANT OPTION']
    # with_option ['GRANT OPTION', 'MAX_QUERIES_PER_HOUR 60',
    #   'MAX_UPDATES_PER_HOUR 75', 'MAX_CONNECTIONS_PER_HOUR 90',
    #   'MAX_USER_CONNECTIONS 5']
    action :grant
    only_if { secret['mysqladmin'] }
  end # mysql_database_user


  # grant privileges to 'insql' for internal, read/write access to The Matrix
  mysql_database_user 'insql' do
    connection    connection_info
    password      secret['insql'] || 'missing_password'
    host          domain
    database_name 'matrix_production'
    privileges    %w{ SELECT INSERT UPDATE }
    action        :grant
    only_if { secret['insql'] }
  end # mysql_database_user
end # .each


# use fetch method to fail if key is missing
node['rails_app']['stages'].each do |stage|
  # create <stage> database
  mysql_database stage.fetch('db_database') do
    connection  connection_info
    encoding    stage.fetch('db_encoding')
    collation   stage.fetch('db_collation')
  end # mysql_database

  # grant privileges to <db_username> for Rails <stage> environment
  mysql_database_user "#{stage}_#{stage.fetch('db_username')}" do
    connection    connection_info
    password      secret[stage.fetch('db_username')] || 'missing_password'
    database_name stage.fetch('db_database')
    username      stage.fetch('db_username')
    host          stage.fetch('db_host')
    action        :grant
    only_if { secret[stage.fetch('db_username')] }
  end # mysql_database_user

end # .each

# # create production, staging databases
# %w(matrix_production matrix_staging).each do |database|
#   mysql_database database do
#     connection connection_info
#     encoding 'utf8'
#     collation 'utf8_general_ci'
#   end # mysql_database database
# end # .each


# # grant privileges to 'matrix' for Rails production environment
# mysql_database_user 'matrix' do
#   connection connection_info
#   password secret['matrix']
#   database_name 'matrix_production'
#   host 'localhost'
#   action :grant
#   only_if { secret['matrix'] }
# end # mysql_database_user


# # grant privileges to 'matrix_staging' for Rails staging environment
# mysql_database_user 'matrix_staging' do
#   connection connection_info
#   password secret['matrix_staging']
#   database_name 'matrix_staging'
#   host 'localhost'
#   action :grant
#   only_if { secret['matrix_staging'] }
# end # mysql_database_user


# grant privileges to 'wwuser' for internal, read-only access to The Matrix
mysql_database_user 'wwuser' do
  connection    connection_info
  password      secret['wwuser'] || 'missing_password'
  database_name 'matrix_production'
  host          '%'
  privileges    ['SELECT']
  action        :grant
  only_if { secret['wwuser'] }
end # mysql_database_user


# drop test database
mysql_database 'test' do
  connection  connection_info
  action      :drop
end # mysql_database


attributes/default.rb
# coding: utf-8
#
# Cookbook Name:: rails_app
# Attributes:: default
#

default['rails_app']['name']                  = 'matrix'
default['rails_app']['rails']['version']      = '3.2.11'
default['rails_app']['stages'] = [
  {
    'name'                    => 'production',
    'codename'                => 'neo',
    'passenger_min_instances' => 8,
    'db_port'                 => 3306,
    'db_database'             => 'matrix_production',
    'db_username'             => 'matrix',
    'db_encoding'             => 'utf8',
    'db_collation'            => 'utf8_general_ci',
    'db_host'                 => 'localhost'
  },
  {
    'name'                    => 'staging',
    'codename'                => 'morpheus',
    'passenger_min_instances' => 1,
    'db_port'                 => 3306,
    'db_database'             => 'matrix_staging',
    'db_username'             => 'matrix_staging',
    'db_encoding'             => 'utf8',
    'db_collation'            => 'utf8_general_ci',
    'db_host'                 => 'localhost'
  }
]


recipes/web.rb
# coding: utf-8
#
# Cookbook Name:: rails_app
# Recipe:: web
#
# Author:: Doc Walker (<doc.walker@jameshardie.com>)
#
# Copyright 2012-2013, James Hardie Building Products, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

include_recipe 'chef-sugar'
include_recipe 'yum::epel' if platform_family?('rhel')

# retrieve contents of encrypted data bag (refer to chef-repo/ENCRYPTED.md)
secret = encrypted_data_bag_item(:encrypted, node.chef_environment)

#-------------------------------------------------------- install dependencies
Chef::Config['yum_timeout'] = 1800
# gem ruby-odbc-0.99994 requires unixODBC-devel
package 'unixODBC-devel'

# gem tiny_tds-0.5.1 requires freetds, freetds-devel
%w(freetds freetds-devel).each do |pkg|
  package pkg
end # .each

# python 2.6 is required; CentOS 5.x needs to use python26 package
if platform_family?('rhel') && node['platform_version'].to_f < 6.0
  package 'python26'
else
  package 'python'
end # if

# nodejs requires python 2.6; the symlink below allows #!/usr/bin/env python
# to use v2.6, provided /usr/local/bin comes before /usr/bin in $PATH
# action :create allows it to repair missing link
link '/usr/local/bin/python' do
  to    '/usr/bin/python26'
  owner 'root'
  group 'root'
  action :create
  only_if { platform_family?('rhel') && node['platform_version'].to_f < 6.0 }
end # link

#------------------ create user, directories, symlinks, database configuration
# user must exist in order to adjust owner/group settings below
user 'jeeves'

# create base directory for public html access
directory '/var/www/html' do
  owner     'root'
  group     'root'
  mode      '0755'
  recursive true
end # directory

node['rails_app']['stages'].each do |stage|

  # ln -s ../apps/<app>/<stage>/current/public /var/www/html/<codename>
  link "/var/www/html/#{stage['codename']}" do
    to    "../apps/#{node['rails_app']['name']}/#{stage['name']}/" +
      'current/public'
    owner 'root'
    group 'root'
  end # link

  base = "/var/www/apps/#{node['rails_app']['name']}/#{stage['name']}"
  # create /var/www/apps/<app>/<stage>/shared/config directory
  directory "#{base}/shared/config" do
    owner     'jeeves'
    group     'jeeves'
    mode      '0755'
    recursive true
  end # directory

  # ln -s /tmp /var/www/apps/<app>/<stage>/current
  link "#{base}/current" do
    to    '/tmp'
    owner 'jeeves'
    group 'jeeves'
    not_if { ::File.symlink?("#{base}/current") }
  end # link

  file "#{base}/current/REVISION" do
    owner   'jeeves'
    group   'jeeves'
    mode    '0755'
    action  :create_if_missing
  end # file

  # create rails database configuration file
  template 'database.yml' do |t|
    path      "#{base}/shared/config/database.yml"
    owner     'jeeves'
    group     'jeeves'
    mode      '0600'
    variables(
      :stage        => stage['name'],
      :db_port      => stage['db_port'],
      :db_database  => stage['db_database'],
      :db_username  => stage['db_username'],
      :db_password  => secret[stage['db_username']]
    )
  end # template

end # .each

# adjust permissions on /var/www/html/<app>
execute "chown -R jeeves:jeeves /var/www/apps/#{node['rails_app']['name']}"

#------------------------------------------------------- configure logrotate.d
include_recipe 'logrotate'

# create configuration file in /etc/logrotate.d/
logrotate_app 'var_www_apps' do
  cookbook    'logrotate'
  path        '/var/www/apps/**/**/shared/log/*.log'
  frequency   'daily'
  options     %w(missingok compress delaycompress sharedscripts)
  rotate      30
  postrotate  'touch /var/www/apps/**/**/current/tmp/restart.txt'
end # logrotate_app

#----------------------------------------------------- configure github access
directory '/home/jeeves/.ssh' do
  owner   'jeeves'
  group   'jeeves'
  mode    '0700'
end # directory

template '/home/jeeves/.ssh/config' do |t|
  owner   'jeeves'
  group   'jeeves'
  mode    '0644'
  variables(
    :header => node['file']['header'].gsub('@filename', t.name)
      .gsub('@hostname', node['hostname'])
  )
end # template

template '/home/jeeves/.ssh/matrix_deploy_key' do |t|
  source  'ssh_key.erb'
  owner   'jeeves'
  group   'jeeves'
  mode    '0600'
  variables(
    :header => node['file']['header'].gsub('@filename', t.name)
      .gsub('@hostname', node['hostname']),
    :private_key => secret['matrix_deploy_key']
  )
end # template
